pkg/helmreconciler/listeners.go

```diff
@@ -0,0 +1,472 @@
+}
+
+// BeginDelete delegates BeginDelete to the Listeners in first to last order.
+func (l *CompositeRenderingListener) BeginDelete(instance runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.BeginDelete(instance); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// BeginChart delegates BeginChart to the Listeners in first to last order.
+func (l *CompositeRenderingListener) BeginChart(chart string, manifests []manifest.Manifest) ([]manifest.Manifest, error) {
+	allErrors := make([]error, 0, len(l.Listeners))
+	var err error
+	for _, listener := range l.Listeners {
+		if manifests, err = listener.BeginChart(chart, manifests); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return manifests, utilerrors.NewAggregate(allErrors)
+}
+
+// BeginResource delegates BeginResource to the Listeners in first to last order.
+func (l *CompositeRenderingListener) BeginResource(obj runtime.Object) (runtime.Object, error) {
+	allErrors := make([]error, 0, len(l.Listeners))
+	var err error
+	for _, listener := range l.Listeners {
+		if obj, err = listener.BeginResource(obj); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return obj, utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceCreated delegates ResourceCreated to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceCreated(created runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceCreated(created); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceUpdated delegates ResourceUpdated to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceUpdated(updated runtime.Object, old runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceUpdated(updated, old); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceDeleted delegates ResourceDeleted to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceDeleted(deleted runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceDeleted(deleted); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceError delegates ResourceError to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceError(obj runtime.Object, err error) error {
+	// reverse order for completions
+	allErrors := make([]error, 0, len(l.Listeners))
+	for index := len(l.Listeners) - 1; index > -1; index-- {
```

Comment:
for l, idx := range l.Listeners
here and elsewhere

Reply:
this is reverse order iteration.